cql3: improve support for C-style parenthesis casts

cvybhu opened the pull request:
CQL supports type casting using C-style casts.
For example it's possible to do: `blob_column = (blob)funcReturningInt()`

This functionality is pretty limited, we only allow such casts between types that have a compatible binary representation. Compatible means that the bytes will stay unchanged after the conversion.
This means that it's legal to cast an int to blob (int is just a 4 byte blob), but it's illegal to cast a bigint to int (change 4 bytes -> 8 bytes).
This simplifies things, to cast we can just reinterpret the value as the other type.

Another use of C-style casts are type hints. Sometimes it's impossible to infer the exact type of an expression from the context. In such cases the type can be specified by casting the expression to this type.
For example: `overloadedFunction((int)?)`
Without the cast it would be impossible to guess what should be the bind marker's type. The function is overloaded, so there are many possible argument types. The type hint specifies that the bind marker has type int.

An interesting thing is that such casts don't have to be explicit. CQL allows to put an int value in a place where a blob value is expected and it will be automatically converted without any explicit casting.

---

I started looking at our implementation of casts because of #12900. In there the author expressed the need to specify a type hint for bind marker used to pass the WASM code. It could be either `(text)?` for text WASM, or `(blob)?` for binary WASM. This specific use of type hints wasn't supported because there was no `receiver` and the implementation of `prepare_expression` didn't handle that. Preparing casts without a receiver should be easy to implement - we can infer the type of the expression by looking at the type to which the expression is cast.

But while reading `prepare_expression` for `expr::cast` I noticed that the code there is a bit strange. The implementation prepared the expression to cast using the original `receiver` instead of a receiver with the cast type. This caused some issues because of which casting didn't work as expected.
For example it was possible to do:
```cql
blob_column = (blob)funcReturningInt()
```
But this didn't work at all:
```cql
blob_column = (blob)(int)12323
```
It tried to prepare `untyped_contant(12323)` with a `blob` receiver, which fails.

This makes `expr::cast` useless for casting. Casting when the representation is compatible is already implicit. I couldn't find a single case where adding a cast would change the behavior in any way.
There was some use for it as a type hint to choose a specific overload of a function, but it was worthless for casting.

Cassandra has the same issue, I created a `cql-pytest` test and it showed that we behave in the same way as Cassandra does.

I decided to improve this. By preparing the expression using a receiver with the cast type, `expr::cast` becomes actually useful for casting values. Things like `(blob)(int)12323` now work without any issues.
This diverges from the behavior in Cassandra, but it's an extension, not a breaking incompatibility.

---

This PR improves `prepare_expression` for `expr::cast` in the following ways:
1) Support for more complex casts by preparing the expression using a different receiver. This makes casts like `(blob)(int)123` possible
2) Support preparing `expr::cast` without a receiver. Type inference chooses the cast type as the type of the expression.
3) Add pytest tests for C-style casts


`2)` Is needed for #12900, the other changes is just something I decided to do since I was already working on this piece of code.

avikivity commented on test/cql-pytest/test_cast.py:
```diff
@@ -0,0 +1,165 @@
+    # Putting (int) before the integer specifies the exact type and it's possible to cast the value.
+    # This type of assignment isn't currently allowed, will work in the future.
+    with pytest.raises(InvalidRequest, match='blob'):
+        cql.execute(f"INSERT INTO {table1} (pk, blob_col) VALUES ({pk}, (int)1234)")
```
I don't think it should work.

cvybhu replied:
Such implicit casts are generally allowed, but I opened #13054 based on our previous discussion.
Disallowing implicit casts would be a breaking change. I guess we could allow it only in cases that worked previously, but it would be inconsistent and a nightmare to maintain the behavior for all the particular cases.

avikivity replied:
Can you give some examples where implicit casts are done and not done currently?

The idea is, if we implicitly cast in many places and the change opens up a few more, it's not a big deal.

But if we implicitly cast in just a few places and the change opens it up to many more places, then we're worsening type checking significantly.

avikivity replied:
btw, something like (blob)(int)1234 can work.

avikivity commented on cql3/expr/prepare_expr.cc:
```diff
@@ -731,26 +731,31 @@ bind_variable_prepare_expression(const bind_variable& bv, data_dictionary::datab
+        [&](const shared_ptr<cql3_type::raw>& raw_type) { return raw_type->prepare(db, keyspace).get_type(); },
+        [](const data_type& prepared_type) {return prepared_type;}
+    }, c.type);
+
```
This also allows us to inject artificial casts, if we want.

avikivity commented on cql3/expr/prepare_expr.cc:
```diff
@@ -780,8 +780,9 @@ cast_prepare_expression(const cast& c, data_dictionary::database db, const sstri
-        return std::nullopt;
+        sstring receiver_name = format("cast({}){:user}", cast_type->cql3_type_name(), c.arg);
+        receiver = make_lw_shared<column_specification>(
+            keyspace, "unkown_cf", ::make_shared<column_identifier>(receiver_name, true), cast_type);
```
spelling

avikivity commented on test/boost/expr_test.cc:
```diff
@@ -1394,6 +1394,31 @@ BOOST_AUTO_TEST_CASE(prepare_cast_text_int) {
+
+    BOOST_REQUIRE(expr::type_of(prepared) == utf8_type);
+}
+
```
Please also add cases that do have a receiver. Both working cases (receiver is also text) and non-working cases.

Corresponding to `UPDATE ... SET text_col = (text) ?` and `UPDATE ... SET int_col = (text) ?`

cvybhu replied:
Added test cases for:
```
text_col = (text)?
int_col = (int)?
text_col = (int)?
int_col = (text)?
```

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/4684/

nyh commented on test/cql-pytest/test_cast.py:
```diff
@@ -47,7 +47,7 @@ def table2(cql, test_keyspace):
 
 # Test casting int values to blob values using C-style casts
-def test_cast_int_blob(cql, table1):
+def test_cast_int_blob(cql, table1, scylla_only):
```
In the first patch I suggested that you split this long test functions into several short test functions. Then you can mark individual tests scylla_only while the others remain true also on Cassandra. At a minimum you should split this into two tests - one scylla_only and the other one not - but I suggest splitting into more than two.

In general, as much as the test code as possible should not have scylla_only, otherwise it is impossible to verify that the test doesn't just enshrine a bug.

nyh reviewed (changes requested):
Very nice. I just left some minor requests regarding the testing.

cvybhu commented on test/cql-pytest/test_cast.py:
```diff
@@ -0,0 +1,165 @@
+    with pytest.raises(InvalidRequest, match='blob'):
+        cql.execute(f"INSERT INTO {table1} (pk, blob_col) VALUES ({pk}, 12)")
+
+    # Putting (blob) before the integer also fails, it's still unkown what the exact type is.
```
Fixed, damn I need to find myself a spellchecker

cvybhu commented on test/cql-pytest/test_cast.py:
```diff
@@ -0,0 +1,165 @@
+    # This type of assignment isn't currently allowed, will work in the future.
+    with pytest.raises(InvalidRequest, match='blob'):
+        cql.execute(f"INSERT INTO {table1} (pk, blob_col) VALUES ({pk}, (int)1234)")
+        assert list(cql.execute(f"SELECT pk, blob_col FROM {table1} WHERE pk = {pk}")) == [(pk, int(1234).to_bytes(4, 'big'))]
```
I did it this way because I thought that it makes it a bit easier to review.
The first time these tests are introduced the reviewer can read all of the tests in the file and decide that they seem to make some sense.
Then when the tests are modified by removing `pytest.raises` there are no other changes - all of the mental load of reading the tests is during the first read, the second time around the change only removes `pytest.raises`, so there is no need for a full context switch to test reading for the reviewer.
On the other hand it might be hard to believe that a line is correct when it never gets executed.

I have no strong opinion, changed it.

cvybhu commented on test/cql-pytest/test_cast.py:
```diff
@@ -0,0 +1,165 @@
+
+    # Converting an int to blob is allowed, but converting a blob to int isn't.
+    # An int can always be converted to a valid blob, but blobs might have wrong amount of bytes
+    # and can't be converted to a valid int.
```
Good idea, I split it into a bunch of smaller functions.

cvybhu commented:
v2:
* Fixed spelling typos
* Added more boost tests for preparing with a receiver, for the following cases: `int_col = (int)?`, `text_col = (text)?`, `int_col = (text)?`, `text_col = (int)?`
* Split the pytest test into multiple smaller tests, now only a part of the tests has to be marked `scylla_only`
* Removed lines that aren't executed from the initial pytest tests

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/4798/

nyh commented on test/cql-pytest/test_cast.py:
```diff
@@ -49,7 +49,7 @@ def table2(cql, test_keyspace):
 # by assigning the smallest type that can fit the constant, but I think that we shouldn't allow converting them
 # to blob without explicitly specifying size of the integer.
-def test_cast_int_literal_to_blob(cql, table1):
+def test_cast_int_literal_to_blob(cql, table1, scylla_only):
```
I think you added scylla_only to this test by mistake?

cvybhu replied:
Right, it's not needed here, fixed.

nyh commented on test/cql-pytest/test_cast.py:
```diff
@@ -61,11 +61,10 @@ def test_cast_int_literal_to_blob_with_blob_cast(cql, table1):
 # Putting (int) before the integer specifies the exact type and it's possible to cast the value.
-# This type of assignment isn't currently allowed, will work in the future.
-def test_cast_int_literal_with_type_hint_to_blob(cql, table1):
+def test_cast_int_literal_with_type_hint_to_blob(cql, table1, scylla_only):
```
It would be nice to have a comment why the test is scylla_only. Since most of the new scylla_only tests have the same explanation on why they are scylla_only, I think it's enough to have this explanation just once in this file.

nyh reviewed (changes requested):
I think there's one misplaced scylla_only on one test, but other than that, looks great.

avikivity commented on test/boost/expr_test.cc:
```diff
@@ -1419,6 +1419,77 @@ BOOST_AUTO_TEST_CASE(prepare_cast_bind_var_no_receiver) {
+        cast{.arg = bind_variable{.bind_index = 0, .receiver = nullptr}, .type = cql3_type::raw::from(utf8_type)};
+
+    BOOST_REQUIRE_THROW(prepare_expression(cast_expr, db, "test_ks", table_schema.get(), make_receiver(int32_type)),
+                        exceptions::invalid_request_exception);
```
This throws, as I want but not as I expect from previous discussion. Ah, it's only blob types that we can decay into. Sort of like void* in C.

cvybhu replied:
Yes, when casting an int to blob we can be sure that we can just take the 4 bytes and reinterpret them as a blob.
With `int` and `text` it's not that simple - not every `int` is a valid `text`, but also it doesn't make much sense to cast `text` to `int`.
I guess we treat `blobs` as a bit special, since it makes *some* sense to cast anything to `blob`.

avikivity commented:
Looks okay, but I'm still confused about the implicit cast into blobs.

cvybhu commented:
v3:
* Removed needless `scylla_only` on `test_cast_int_literal_to_blob`
* Added a note why some of the tests require scylla_only, it's at the top of the file.

cvybhu commented:
> Can you give some examples where implicit casts are done and not done currently?
>
> The idea is, if we implicitly cast in many places and the change opens up a few more, it's not a big deal.
> 
> But if we implicitly cast in just a few places and the change opens it up to many more places, then we're worsening type checking significantly.

I don't know of any places that currently use such casts. I guess it would be possible to find some by banning them and seeing what breaks.

I agree that it would be better if such implicit casts weren't allowed. I'm all for super strict type systems wherever possible,  I even like how Rust forces me to think that I'm now casting `int64` to `int32` as there might be an overflow. Strict type checking saved me more than once.

Here I decided to go forward with them for a few reasons:
* It's something that Cassandra supports. We want our data model to be as compatible as possible with Cassandra, so if Cassandra allows such casts then so should we. Looking around the expression code there's `test_assignment` everywhere which clearly intended to support such casts.
* The current implementation of these casts doesn't make much sense to me, it seems inconsistent. Why sometimes casting an int to blob is legal, and sometimes it's illegal? Why is an `int` different when returned from a function than when written like `(int)123`?.
* Trying to maintain compatibility with Cassandra and previous versions of Scylla is very hard when the implementation is so inconsistent. In each change we would have to think hard to make sure that it still allows the casts in random places, and disallows them in other random places. Having one consistent way the casts work greatly simplifies maintenance.
* I want to have a general rule to check `receiver` correctness in `prepare_expression`. Right now for each type of expression we have randomly either direct type comparison or `test_assignment`. I want to check receiver type once for all the types, but this requires a consistent vision of what's right.
* Overall I think it makes more sense to have a consistent rule for casting in CQL, it reduces the chaos there. It's better than random support or not, no half measures.

> But if we implicitly cast in just a few places and the change opens it up to many more places, then we're worsening type checking significantly.

I don't think we're opening it up in that many  places. Doing `blob_col = ?` still expects the `?` to be a blob, `blob_col = 123` won't work. Other things, like function return values and arguments already allowed implicit casts before my change.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/4856/
